lib/src/ignore.dart

```diff
@@ -0,0 +1,560 @@
+      );
+    }
+    if (_rules.isEmpty) {
+      return false;
+    }
+    final pathWithoutSlash =
+        path.endsWith('/') ? path.substring(0, path.length - 1) : path;
+    return unignoredFiles(
+      beneath: pathWithoutSlash,
+      includeDirs: true,
+      listDir: (dir) {
+        // List the next part of path:
+        if (dir == pathWithoutSlash) return [];
+        final startOfNext = dir.isEmpty ? 0 : dir.length + 1;
+        final nextSlash = path.indexOf('/', startOfNext);
+        return [path.substring(startOfNext, nextSlash)];
+      },
+      ignoreForDir: (dir) => dir == '' ? this : null,
+      isDir: (candidate) =>
+          path.length > candidate.length && path[candidate.length] == '/',
+    ).isEmpty;
+  }
+
+  /// Returns all the files in the tree under (and including) [beneath] not
+  /// ignored by ignore-files from [root] and down.
+  ///
+  /// Represents paths normalized  using '/' as directory separator. The empty
+  /// relative path is '.', no '..' are allowed.
+  ///
+  /// [beneath] must start with [root] and even if it is a directory it should not
+  /// end with '/', if [beneath] is not provided, everything under root is
+  /// included.
+  ///
+  /// [listDir] should enumerate the immediate contents of a given directory,
+  /// returning paths including [root].
+  ///
+  /// [isDir] should return true if the argument is a directory. It will only be
+  /// queried with file-names under (and including) [beneath]
+  ///
+  /// [ignoreForDir] should retrieve the ignore rules for a single directory
+  /// or return `null` if there is no ignore rules.
+  ///
+  /// If [includeDirs] is true non-ignored directories will be included in the
+  /// result (including beneath).
+  ///
+  /// This example program lists all files under second argument that are
+  /// not ignored by .gitignore files from first argument and below:
+  ///
+  /// ```dart
+  /// import 'dart:io';
+  /// import 'package:path/path.dart' as p;
+  /// import 'package:pub/src/ignore.dart';
+  ///
+  /// void main(List<String> args) {
+  ///   var root = p.normalize(args[0]);
+  ///   if (root == '.') root = '';
+  ///   var beneath = args.length > 1 ? p.normalize(args[1]) : root;
+  ///   if (beneath == '.') beneath = '';
+  ///   String resolve(String path) {
+  ///     return p.joinAll([root, ...p.posix.split(path)]);
+  ///   }
+  ///
+  ///   Ignore.unignoredFiles(
+  ///     beneath: beneath,
+  ///     listDir: (dir) => Directory(resolve(dir)).listSync().map((x) {
+  ///        final relative = p.relative(x.path, from: root);
+  ///       return p.posix.joinAll(p.split(relative));
+  ///     }),
+  ///     ignoreForDir: (dir) {
+  ///       final f = File(resolve('dir/.gitignore'));
+  ///       return f.existsSync() ? Ignore([f.readAsStringSync()]) : null;
+  ///     },
+  ///     isDir: (dir) => Directory(resolve(dir)).existsSync(),
+  ///   ).forEach(print);
+  /// }
+  /// ```
+  static List<String> unignoredFiles({
```

Comment:
unignored implies that there is a negative ignore pattern, doesn't it?

Why about just `listFiles`? or `listFilesNotIgnored` or `filesNotIgnored`..